src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -244,18 +273,96 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() {
+  //       enough runway to recalibrate our "average allocation rate".
+  //
+  //    3. Though there is an "average" allocation rate, a given workload's demand for allocation may be very bursty.  We
+  //       allocate a bunch of LABs during the 5 ms that follow completion of a GC, then we perform no more allocations for
+  //       the next 150 ms.  It seems we want the "spike" to represent the maximum divergence from average within the
+  //       period of time between consecutive evaluation of the should_start_gc() service.  Here's the thinking:
+  //
+  //       a) Between now and the next time I ask whether should_start_gc(), we might experience a spike representing
+  //          the anticipated burst of allocations.  If that would put us over budget, then we should start GC immediately.
+  //       b) Between now and the anticipated depletion of allocation pool, there may be two or more bursts of allocations.
+  //          If there are more than one of these bursts, we can "approximate" that these will be separated by spans of
+  //          time with very little or no allocations so the "average" allocation rate should be a suitable approximation
+  //          of how this will behave.
+  //
+  //    For cases 1 and 2, we need to "quickly" recalibrate the average allocation rate whenever we detect a change
+  //    in operation mode.  We want some way to decide that the average rate has changed.  Make average allocation rate
+  //    computations an independent effort.
+
   size_t spike_headroom = capacity / 100 * ShenandoahAllocSpikeFactor;
   size_t penalties      = capacity / 100 * _gc_time_penalties;
 
+  // TODO: Account for inherent delays in responding to GC triggers
+  //  1. It has been observed that delays of 200 ms or greater are common between the moment we return true from should_start_gc()
+  //     and the moment at which we begin execution of the concurrent reset phase.  Add this time into the calculation of
+  //     avg_cycle_time below.  (What is "this time"?  Perhaps we should remember recent history of this delay for the
+  //     running workload and use the maximum delay recently seen for "this time".)
+  //  2. The frequency of inquiries to should_start_gc() is adaptive, ranging between ShenandoahControlIntervalMin and
+  //     ShenandoahControlIntervalMax.  The current control interval (or the max control interval) should also be added into
+  //     the calculation of avg_cycle_time below.
+
   allocation_headroom -= MIN2(allocation_headroom, spike_headroom);
   allocation_headroom -= MIN2(allocation_headroom, penalties);
 
   double avg_cycle_time = _gc_time_history->davg() + (_margin_of_error_sd * _gc_time_history->dsd());
+
+  size_t last_live_memory = get_last_live_memory();
+  size_t penultimate_live_memory = get_penultimate_live_memory();
+  double original_cycle_time = avg_cycle_time;
+  if ((penultimate_live_memory < last_live_memory) && (penultimate_live_memory != 0)) {
```

Comment:
The heuristic uses an exponentially weighted average for cycle times. The decay factor can be adjusted from the command line. The margin of error for estimating cycle times is also adjusted in response to degenerated cycles or successful concurrent cycles. Perhaps we should have the heuristic also adjust the decay factor? Have we evaluated the results of these changes on a variety of work loads?

Reply:
Exposure to variety of workload conditions is very limited.  All of our work on heuristics needs much more attention.  Maybe we don't want to fiddle with these "quick and dirty" fixes until after we can do more thorough experimentation with different workloads.

One important observation I've made is that being too aggressive with heuristics is more harmful with generational mode than traditional mode.  With traditional mode, being too aggressive means we spend a few more cycles in GC.  With generational mode, it might mean we "starve" old-gen GC, and that might result in degenerated or Full STW pauses.

I'm ok to pull out these heuristic tweaks.  I expect this will all be re-done eventually.  I'm not sure how to deliver the best out-of-box experience to early evaluators.